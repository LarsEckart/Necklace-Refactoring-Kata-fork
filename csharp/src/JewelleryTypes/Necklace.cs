using NecklaceRefactoringKata.Enums;

namespace NecklaceRefactoringKata.JewelleryTypes
{
    public record Necklace(NecklaceType Type, Jewel Stone) : JewelleryBase(Stone)
    {
    }
}
