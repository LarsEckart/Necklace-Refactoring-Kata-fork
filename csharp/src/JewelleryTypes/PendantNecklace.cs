using NecklaceRefactoringKata.Enums;

namespace NecklaceRefactoringKata.JewelleryTypes
{
    public record PendantNecklace(NecklaceType Type, Jewel Stone, Necklace Chain, JewelleryBase Pendant) : Necklace(Type, Stone)
    {
        public override bool IsLarge()
        {
            return Chain.IsLarge() || Pendant.IsLarge();
        }
    }
}
