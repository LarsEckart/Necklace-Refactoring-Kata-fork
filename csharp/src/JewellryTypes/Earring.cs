using NecklaceRefactoringKata.Enums;

namespace NecklaceRefactoringKata.Models
{
    public record Earring(EarringType Type, Jewel Stone) : JewelleryBase(Stone)
    {
        public override bool IsSmall()
        {
            return Type == EarringType.Stud;
        }
        public override bool IsEarring()
        {
            return true;
        }
    }
}
