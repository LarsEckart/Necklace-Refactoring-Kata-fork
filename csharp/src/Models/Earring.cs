using NecklaceRefactoringKata.Enums;

namespace NecklaceRefactoringKata.Models
{
    public record Earring(EarringType Type, Jewel Stone) : Jewellery(Stone, IsSmall: Type == EarringType.Stud, IsEarring: true);
}
