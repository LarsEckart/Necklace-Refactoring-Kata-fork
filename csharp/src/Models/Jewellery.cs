using NecklaceRefactoringKata.Enums;

namespace NecklaceRefactoringKata.Models
{
    public abstract record Jewellery(Jewel Stone)
    {
        public virtual bool IsRing() { return false; }
        public virtual bool IsSmall() { return false; }
        public virtual bool IsEarring() { return false; }
        public virtual bool IsNecklace() { return false; }
        public virtual bool IsLarge() { return false; }
    }
}
