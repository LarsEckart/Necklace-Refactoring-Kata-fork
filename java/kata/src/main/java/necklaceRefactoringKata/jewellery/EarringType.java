package necklaceRefactoringKata.jewellery;

public enum EarringType {
    Stud,
    Hoop,
    Drop
}
