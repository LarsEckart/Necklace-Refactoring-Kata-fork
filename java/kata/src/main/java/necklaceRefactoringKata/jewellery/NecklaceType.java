package necklaceRefactoringKata.jewellery;

public enum NecklaceType {
    Beads,
    Chain,
    Pendant,
    LongChain
}
