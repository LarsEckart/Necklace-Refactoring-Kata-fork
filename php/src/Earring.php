<?php

namespace NecklaceRefactoringKata;

enum EarringType {
    case Stud;
    case Hoop;
    case Drop;
}

class Earring extends Jewellery
{
    public function __construct(private EarringType $earringType)
    {
    }

    public function isSmall(): bool
    {
        return $this->earringType === EarringType::Stud;
    }

    public function isEarring(): bool
    {
        return true;
    }
}

