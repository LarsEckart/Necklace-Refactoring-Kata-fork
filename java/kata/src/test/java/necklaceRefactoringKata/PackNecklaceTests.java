package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Earring;
import necklaceRefactoringKata.jewellery.EarringType;
import necklaceRefactoringKata.jewellery.Jewel;
import necklaceRefactoringKata.jewellery.Jewellery;
import necklaceRefactoringKata.jewellery.Necklace;
import necklaceRefactoringKata.jewellery.NecklaceType;
import necklaceRefactoringKata.jewellery.Pendant;
import necklaceRefactoringKata.jewellery.PendantNecklace;
import necklaceRefactoringKata.jewellery.Ring;
import org.approvaltests.combinations.CombinationApprovals;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;

class PackNecklaceTests {

    @Test
    void test_necklace() {
        CombinationApprovals.verifyAllCombinations(this::necklaces, Jewel.values(), NecklaceType.values());
    }

    @Test
    void test_earring() {
        CombinationApprovals.verifyAllCombinations(this::earrings, Jewel.values(), EarringType.values());
    }

    @Test
    void test_pendant() {
        CombinationApprovals.verifyAllCombinations(this::pendants, Jewel.values());
    }

    @Test
    void test_pendantNecklace() {
        CombinationApprovals.verifyAllCombinations(this::pendantNecklaces, Jewel.values(), NecklaceType.values(), allNecklaces(), allJewellery());
    }

    private Necklace[] allNecklaces() {
        var result = new ArrayList<>();
        for (Jewel jewel : Jewel.values()) {
            for (NecklaceType necklaceType : NecklaceType.values()) {
                result.add(new Necklace(jewel, necklaceType));
            }
        }
        return result.toArray(new Necklace[0]);
    }

    private Jewellery[] allJewellery() {
        var result = new ArrayList<>();
        for (Jewel jewel : Jewel.values()) {
            result.add(new Ring(jewel));
            result.add(new Pendant(jewel));
            for (NecklaceType necklaceType : NecklaceType.values()) {
                result.add(new Necklace(jewel, necklaceType));
            }
            for (EarringType earringType : EarringType.values()) {
                result.add(new Earring(jewel, earringType));
            }
        }
        return result.toArray(new Jewellery[0]);
    }

    @Test
    void test_ring() {
        CombinationApprovals.verifyAllCombinations(this::rings, Jewel.values());
    }

    private String necklaces(Jewel jewel, NecklaceType type) {
        var item = new Necklace(jewel, type);
        var storage = new JewelleryStorage();

        Packer.packNecklace(item, storage);

        return StoragePrinter.print(storage);
    }

    private String earrings(Jewel jewel, EarringType type) {
        var item = new Earring(jewel, type);
        var storage = new JewelleryStorage();

        Packer.pack(item, storage);

        return StoragePrinter.print(storage);
    }

    private String pendants(Jewel jewel) {
        var item = new Pendant(jewel);
        var storage = new JewelleryStorage();

        Packer.pack(item, storage);

        return StoragePrinter.print(storage);
    }

    private String pendantNecklaces(Jewel jewel, NecklaceType type, Necklace chain, Jewellery pendant) {
        var item = new PendantNecklace(jewel, type, chain, pendant);
        var storage = new JewelleryStorage();

        Packer.pack(item, storage);

        return StoragePrinter.print(storage);
    }

    private String rings(Jewel jewel) {
        var item = new Ring(jewel);
        var storage = new JewelleryStorage();

        Packer.pack(item, storage);

        return StoragePrinter.print(storage);
    }

}
