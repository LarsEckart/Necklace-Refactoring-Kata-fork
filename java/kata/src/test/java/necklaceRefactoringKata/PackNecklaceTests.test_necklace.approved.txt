[Plain, Beads] => tree 
[Plain, Chain] => box top shelf 
[Plain, Pendant] => box top shelf 
[Plain, LongChain] =>  
[Diamond, Beads] => safe 
[Diamond, Chain] => safe 
[Diamond, Pendant] => safe 
[Diamond, LongChain] => safe 
[Pearl, Beads] => tree 
[Pearl, Chain] => box top shelf 
[Pearl, Pendant] => box top shelf 
[Pearl, LongChain] =>  
[Amber, Beads] => tree 
[Amber, Chain] => box top shelf 
[Amber, Pendant] => box top shelf 
[Amber, LongChain] =>  
